### PR TITLE
[Merged by Bors] - Disallow attesting to optimistic head

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Disallow the production of attestations and retrieval of unaggregated attestations when they reference an optimistic head. Add tests to this end.

I also moved `BeaconChain::produce_unaggregated_attestation_for_block` to the `BeaconChainHarness`. It was only being used during tests, so it's nice to stop pretending it's production code. I also needed something that could produce attestations to optimistic blocks in order to simulate scenarios where the justified checkpoint is determined invalid (if no one would attest to an optimistic block, we could never justify it and then flip it to invalid).

## Additional Info

- ~~Blocked on #3126~~

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: I'm sort of thinking we should handle errors here like in `produce_unaggregated_attestation`, otherwise the HTTP API will return "missing aggregate" which would be a bit misleading

**Review by realbigsean — COMMENTED** [review]
Just had one comment, looking good though!

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Great point, I've added that. I also jigged around with the naming too. All the meaningful changes are in [bc8f220](https://github.com/sigp/lighthouse/pull/3140/commits/bc8f220f72653b82ea6d86bb186242542e181878).

**Comment by paulhauner** [comment]
All comments addressed, thanks @realbigsean. I will merge after an approval (or change request) :)

File: consensus/proto_array/src/proto_array_fork_choice.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array_fork_choice.rs]:
```suggestion
    /// - Has a valid payload
```

File: consensus/fork_choice/src/fork_choice.rs
Comment (realbigsean) [inline consensus/fork_choice/src/fork_choice.rs:132]: I'm wondering if `PayloadVerificationStatus::NotVerified` should be renamed to `PayloadVerificationStatus::Optimistic` as well to make this mapping very clear. I could understand if we want to keep the concept of "optimism" constrained to fork choice though. What are your thoughts?

This method can also be used in `block_verification.rs` L1168 by the way

File: beacon_node/http_api/src/lib.rs
Comment (realbigsean) [inline beacon_node/http_api/src/lib.rs]: This error's swallowed 

```suggestion
                        .map_err(|e| {
                            warp_utils::reject::custom_bad_request(format!(
                                "unable to fetch aggregate: {e:?}"
                            ))
                        })?
```

**Review by realbigsean — CHANGES_REQUESTED** [review]

File: consensus/fork_choice/src/fork_choice.rs
Comment (paulhauner) [reply consensus/fork_choice/src/fork_choice.rs:132]: Great suggestions, will do both.

File: beacon_node/http_api/src/lib.rs
Comment (paulhauner) [reply beacon_node/http_api/src/lib.rs]: Nice catch, thank you!

**Comment by paulhauner** [comment]
Thanks for the comments, apologies for the sloppy mistakes!

I've addressed all comments and fixed the merge conflict :)

**Review by realbigsean — APPROVED** [review]
Awesome! LGTM

**Comment by paulhauner** [comment]
I'm blocking this on the v2.2.1 release (#3149)

**Pull request closed without merging**
